File: worker/src/team.js
// team.js — raw ESPN team endpoints → canonical favorite-team shapes. Pure
// functions (no I/O), like normalize.js, so they run in Node tests and the
// worker alike. Reuses normalize.js's per-event builder so team games are
// normalized through the EXACT same path as the scoreboard — no fork.

import { resolve } from '../../schema/tools/resolve.mjs';
import { normalizeScoreboard, buildEvent, https, darkLogoOf } from './normalize.js';

// ---- teams list (the favorites picker) --------------------------------------
// ESPN: { sports:[{ leagues:[{ teams:[{ team:{...} }] }] }] }
export function normalizeTeams(reg, key, raw) {
  const profile = resolve(reg, key);
  const teams = raw?.sports?.[0]?.leagues?.[0]?.teams ?? [];
  return teams
    .map(({ team: t }) => {
      if (!t) return null;
      const light = https(t.logo || t.logos?.[0]?.href);
      const out = {
        id: String(t.id ?? ''),
        displayName: t.displayName || t.name || t.shortDisplayName || '',
        abbreviation: t.abbreviation || undefined,
      };
      if (light) {
        out.logo = light;
        const d = darkLogoOf(t, light, profile.espnSport); // logos[] may carry an explicit dark rel
        if (d) out.logoDark = d;
      }
      if (t.color) out.color = t.color;
      return out;
    })
    .filter(t => t && t.id)
    .sort((a, b) => a.displayName.localeCompare(b.displayName));
}

// ---- shared team identity block ---------------------------------------------
// The card/detail identity, built once from the (schedule.team) block so the
// favorite card and the team-detail page can never fork. `t` is the raw ESPN
// team object; VERIFIED (schedule.team): logo is a STRING (not a logos[] array),
// recordSummary + standingSummary are STRINGS ('46-36', '2nd in AL East').
export function teamIdentityOf(profile, t, teamId) {
  t = t || {};
  const light = https(t.logo || t.logos?.[0]?.href);
  const team = {
    id: String(t.id ?? teamId),
    displayName: t.displayName || t.name || '',
    abbreviation: t.abbreviation || undefined,
    record: t.recordSummary || undefined,
    standingSummary: t.standingSummary || undefined, // absent for national teams
  };
  if (light) {
    team.logo = light;
    const d = darkLogoOf(t, light, profile.espnSport);
    if (d) team.logoDark = d;
  }
  if (t.color) team.color = t.color;
  return team;
}

// ---- team card (live / last / next) -----------------------------------------
// ESPN team schedule: { team:{ id, displayName, abbreviation, color, logo,
// recordSummary, standingSummary, ... }, events:[ <scoreboard-shaped events> ] }.
export function normalizeTeamCard(reg, key, teamId, schedule) {
  const profile = resolve(reg, key);

  const team = teamIdentityOf(profile, schedule?.team, teamId);

  // every scheduled event normalized through the shared builder
  const events = (schedule?.events || []).map(e => buildEvent(profile, e));

  let live = null, last = null, next = null;
  for (const ev of events) {
    const c = ev.competitions[0];
    if (!c) continue;
    const ms = Date.parse(ev.start) || 0;
    const ph = c.status.phase;
    if (ph === 'live') {
      if (!live || (Date.parse(live.start) || 0) > ms) live = ev;     // earliest-started live
    } else if (c.status.ended || ph === 'final') {
      if (!last || (Date.parse(last.start) || 0) < ms) last = ev;     // most-recent ended
    } else if (ph === 'scheduled') {
      if (!next || (Date.parse(next.start) || 0) > ms) next = ev;     // earliest upcoming
    }
    // postponed / canceled / suspended / unknown → not surfaced
  }

  return {
    league: key,
    sport: profile.espnSport,
    leagueName: profile.name || key.split('/')[1] || '',
    team,
    live,
    last,
    next,
    anyLive: live != null,
  };
}

// ---- scoreboard fallback (national teams / tournaments) ----------------------
// ESPN's team-schedule endpoint returns events:[] for national teams and many
// tournament squads (their fixtures live under other ESPN league ids), and a club's
// in-progress game can also lag behind the live scoreboard. So when the schedule
// gave us NO live game, backfill the card from the league scoreboard — the exact
// same canonical slate the Scores feed renders — picking the competition that
// involves this team. Pure: takes the raw scoreboard, returns a patched card.
export function applyScoreboardFallback(reg, key, teamId, card, sb) {
  const norm = normalizeScoreboard(reg, key, sb);
  const id = String(teamId);
  const compFor = ev => ev.competitions.find(c => c.competitors.some(x => x.id === id));
  const mine = norm.events.filter(ev => compFor(ev));
  if (!mine.length) return card;

  let { live, last, next } = card;
  for (const ev of mine) {
    const c = compFor(ev);
    const ms = Date.parse(ev.start) || 0;
    const ph = c.status.phase;
    if (ph === 'live') {
      if (!live || (Date.parse(live.start) || 0) > ms) live = ev;            // earliest-started live
    } else if (c.status.ended || ph === 'final') {
      if (!last || (Date.parse(last.start) || 0) < ms) last = ev;            // most-recent ended
    } else if (ph === 'scheduled') {
      if (!next || (Date.parse(next.start) || 0) > ms) next = ev;            // earliest upcoming
    }
  }

  // Fill team identity from the scoreboard competitor when the (empty) schedule
  // gave us none — otherwise the card header renders blank.
  let team = card.team;
  if (!team.displayName || !team.logo) {
    const ev = live || last || next;
    const me = ev && compFor(ev)?.competitors.find(x => x.id === id);
    if (me) team = {
      ...team,
      displayName: team.displayName || me.displayName,
      abbreviation: team.abbreviation || me.abbreviation,
      logo: team.logo || me.logo,
      logoDark: team.logoDark || me.logoDark,
      color: team.color || me.color,
    };
  }

  return { ...card, team, live, last, next, anyLive: live != null };
}

File: app/lib/src/data/team.dart
// team.dart — Dart port of worker/src/team.js. Raw ESPN team endpoints →
// canonical favorite-team shapes. Reuses the shared per-event builder (buildEvent)
// + normalizeScoreboard so team games normalize through the EXACT same path as the
// scoreboard. Pure.

import 'profiles.dart';
import 'util.dart';
import 'normalize.dart';

/// JS String.prototype.localeCompare (default locale) — case-insensitive primary
/// ordering with a case/codepoint tiebreak. Good enough for team-name sorting;
/// verified against the JS output via the teams golden suite.
int localeCompare(String a, String b) {
  final la = a.toLowerCase(), lb = b.toLowerCase();
  final c = la.compareTo(lb);
  if (c != 0) return c;
  return a.compareTo(b);
}

// ---- teams list (the favorites picker) --------------------------------------
List<Map<String, dynamic>> normalizeTeams(Registry reg, String key, dynamic raw) {
  final profile = resolve(reg, key);
  final teams = field(first(field(first(field(raw, 'sports')), 'leagues')), 'teams');
  final list = teams is List ? teams : const [];
  final out = <Map<String, dynamic>>[];
  for (final entry in list) {
    final t = field(entry, 'team');
    if (t == null) continue;
    final light = https(or([field(t, 'logo'), field(first(field(t, 'logos')), 'href')]));
    final o = pickNN({
      'id': jsStr(field(t, 'id') ?? ''),
      'displayName': or([field(t, 'displayName'), field(t, 'name'), field(t, 'shortDisplayName'), '']),
      'abbreviation': or([field(t, 'abbreviation'), null]),
    }, ['id', 'displayName', 'abbreviation']);
    if (light != null) {
      o['logo'] = light;
      final d = darkLogoOf(t, light, profile['espnSport']);
      if (d != null) o['logoDark'] = d;
    }
    if (truthy(field(t, 'color'))) o['color'] = field(t, 'color');
    if (truthy(o['id'])) out.add(o);
  }
  out.sort((a, b) => localeCompare(a['displayName'] as String, b['displayName'] as String));
  return out;
}

// ---- shared team identity block ---------------------------------------------
Map<String, dynamic> teamIdentityOf(Map profile, dynamic t, dynamic teamId) {
  t ??= {};
  final light = https(or([field(t, 'logo'), field(first(field(t, 'logos')), 'href')]));
  final team = pickNN({
    'id': jsStr(field(t, 'id') ?? teamId),
    'displayName': or([field(t, 'displayName'), field(t, 'name'), '']),
    'abbreviation': or([field(t, 'abbreviation'), null]),
    'record': or([field(t, 'recordSummary'), null]),
    'standingSummary': or([field(t, 'standingSummary'), null]),
  }, ['id', 'displayName', 'abbreviation', 'record', 'standingSummary']);
  if (light != null) {
    team['logo'] = light;
    final d = darkLogoOf(t, light, profile['espnSport']);
    if (d != null) team['logoDark'] = d;
  }
  if (truthy(field(t, 'color'))) team['color'] = field(t, 'color');
  return team;
}

// ---- team card (live / last / next) -----------------------------------------
int _ms(dynamic s) => DateTime.tryParse(jsStr(s))?.millisecondsSinceEpoch ?? 0;

Map<String, dynamic> normalizeTeamCard(Registry reg, String key, dynamic teamId, dynamic schedule) {
  final profile = resolve(reg, key);
  final team = teamIdentityOf(profile, field(schedule, 'team'), teamId);
  final events = (field(schedule, 'events') is List ? field(schedule, 'events') as List : const [])
      .map((e) => buildEvent(profile, e as Map))
      .toList();

  Map<String, dynamic>? live, last, next;
  for (final ev in events) {
    final c = first(ev['competitions']);
    if (c == null) continue;
    final ms = _ms(ev['start']);
    final ph = field(c['status'], 'phase');
    if (ph == 'live') {
      if (live == null || _ms(live['start']) > ms) live = ev;
    } else if (field(c['status'], 'ended') == true || ph == 'final') {
      if (last == null || _ms(last['start']) < ms) last = ev;
    } else if (ph == 'scheduled') {
      if (next == null || _ms(next['start']) > ms) next = ev;
    }
  }

  return {
    'league': key,
    'sport': profile['espnSport'],
    'leagueName': or([profile['name'], key.split('/').length > 1 ? key.split('/')[1] : null, '']),
    'team': team,
    'live': live,
    'last': last,
    'next': next,
    'anyLive': live != null,
  };
}

// ---- scoreboard fallback (national teams / tournaments) ----------------------
Map<String, dynamic> applyScoreboardFallback(Registry reg, String key, dynamic teamId, Map card, Map sb) {
  final norm = normalizeScoreboard(reg, key, sb);
  final id = jsStr(teamId);
  Map? compFor(Map ev) {
    for (final c in ev['competitions'] as List) {
      if ((c['competitors'] as List).any((x) => (x as Map)['id'] == id)) return c as Map;
    }
    return null;
  }

  final mine = (norm['events'] as List).where((ev) => compFor(ev as Map) != null).toList();
  if (mine.isEmpty) return card.cast<String, dynamic>();

  var live = card['live'], last = card['last'], next = card['next'];
  for (final ev in mine) {
    final c = compFor(ev as Map)!;
    final ms = _ms(ev['start']);
    final ph = field(c['status'], 'phase');
    if (ph == 'live') {
      if (live == null || _ms(live['start']) > ms) live = ev;
    } else if (field(c['status'], 'ended') == true || ph == 'final') {
      if (last == null || _ms(last['start']) < ms) last = ev;
    } else if (ph == 'scheduled') {
      if (next == null || _ms(next['start']) > ms) next = ev;
    }
  }

  var team = card['team'] as Map;
  if (!truthy(team['displayName']) || !truthy(team['logo'])) {
    final ev = live ?? last ?? next;
    final me = ev != null ? (compFor(ev as Map)?['competitors'] as List?)?.firstWhere((x) => (x as Map)['id'] == id, orElse: () => null) : null;
    if (me != null) {
      team = {
        ...team,
        'displayName': or([team['displayName'], me['displayName']]),
        'abbreviation': or([team['abbreviation'], me['abbreviation']]),
        'logo': or([team['logo'], me['logo']]),
        'logoDark': or([team['logoDark'], me['logoDark']]),
        'color': or([team['color'], me['color']]),
      };
      // JS spread keeps undefined-valued keys out; drop nulls to match.
      team = pickNN(team, team.keys.map((e) => e.toString()).toList());
    }
  }

  return <String, dynamic>{...card.cast<String, dynamic>(), 'team': team, 'live': live, 'last': last, 'next': next, 'anyLive': live != null};
}
